pubsub/tests/unit/pubsub_v1/publisher/batch/test_thread.py

```diff
@@ -149,6 +151,35 @@ def test_blocking__commit():
     assert futures[1].result() == "b"
 
 
+def test_client_api_publish_not_blocking_additional_publish_calls():
+    batch = create_batch(max_messages=1)
+    api_publish_called = threading.Event()
+
+    def api_publish_delay(_, messages):
+        api_publish_called.set()
+        time.sleep(1.0)
+        message_ids = [str(i) for i in range(len(messages))]
+        return types.PublishResponse(message_ids=message_ids)
+
+    api_publish_patch = mock.patch.object(
+        type(batch.client.api), "publish", side_effect=api_publish_delay
+    )
+
+    with api_publish_patch:
+        batch.publish({"data": b"first message"})
+
+        start = datetime.datetime.now()
+        event_set = api_publish_called.wait(timeout=1.0)
+        if not event_set:
+            pytest.fail("API publish was not called in time")
+        batch.publish({"data": b"second message"})
```

Comment:
Won't this always take 1 second due to the `side_effect` patch?

Reply:
That's the gist of the test - if the lock is held by the batch commit thread while the batch is being committed, i.e. when the code is waiting for `self._client.api.publish()` to return, the operation will take more than one second and the test will fail.

If the lock is released right after the state transition to `IN_PROGRESS`, i.e. before the call to `self._client.api.publish()`, the latter will not block other threads that might (indirectly) call `batch.publish()` in the meantime and try to acquire the state lock.